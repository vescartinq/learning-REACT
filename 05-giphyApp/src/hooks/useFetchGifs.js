// CustomHooks
import { useState, useEffect } from 'react';
import { getGifs } from '../actions/gifActions';

export const useFetchGifs = (category) => {
  // Estado inicial CARGANDO
  const [state, setState] = useState({
    data: [],
    loading: true,
  });

  useEffect(() => {
    getGifs(category).then((images) => {
      setState({
        data: images,
        loading: false,
      });
    });
  }, [category]);

  return state; // Estado inicial: { data:[], loading: true };
};
