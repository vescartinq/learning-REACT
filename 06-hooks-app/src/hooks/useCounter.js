import { useState } from 'react';

export const useCounter = (initialState = 10) => {
  const [counter, setCounter] = useState(initialState); // 10

  const reset = () => {
    setCounter(initialState);
  };

  const increment = (factor) => {
    setCounter(counter + factor);
  };

  const decrement = (factor) => {
    setCounter(counter - factor);
  };

  return {
    counter,
    increment,
    decrement,
    reset,
  };
};
