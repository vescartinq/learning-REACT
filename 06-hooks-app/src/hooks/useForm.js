import { useState } from 'react';

export const useForm = (initialState = {}) => {
  const [values, setValues] = useState(initialState);

  const handleInputChange = (e) => {
    // console.log(e.target.value);

    setValues({
      ...values,
      [e.target.name]: e.target.value,
    });
  };

  return [values, handleInputChange];
};
