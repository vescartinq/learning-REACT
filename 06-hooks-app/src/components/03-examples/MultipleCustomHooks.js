import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

import '../02-useEffect/effects.css';

export const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { author, quote } = !!data && data[0]; //si no es null y no es no data (si existe data), regresa primera posici√≥n de data

  return (
    <div>
      <h1>BreakingBad Quotes</h1>
      <hr />

      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-0"> {quote} </p>
          <footer className="blockquote-footer"> {author} </footer>
        </blockquote>
      )}

      <button className="btn btn-primary" onClick={() => increment(1)}>
        Siguiente quote
      </button>
    </div>
  );
};
