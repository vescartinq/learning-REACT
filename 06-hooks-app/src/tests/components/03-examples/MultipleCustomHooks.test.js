import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
// import { useCounter } from '../../../hooks/useCounter';

jest.mock('../../../hooks/useFetch');
// jest.mock('../../../hooks/useCounter');

describe('Tests <MultipleCustomHooks />', () => {
  //   useCounter.mockReturnValue({
  //     counter: 10,
  //     increment: () => {},
  //   });

  test('should render component', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should show data', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: 'Victor',
          quote: 'Hola Mundo',
        },
      ],
      loading: false,
      error: null,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo');
    expect(wrapper.find('footer').text().trim()).toBe('Victor');
  });
});
