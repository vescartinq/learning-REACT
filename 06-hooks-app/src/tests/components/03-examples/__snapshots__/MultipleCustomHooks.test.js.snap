// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests <MultipleCustomHooks /> should render component 1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</div>
`;
